packages/frontend/src/components/chat.tsx

```diff
@@ -45,37 +52,107 @@ function Message({ msg, from, peerId }: MessageProps) {
 export default function ChatContainer() {
   const { libp2p } = useLibp2pContext()
   const { messageHistory, setMessageHistory } = useChatContext();
+  const { files, setFiles } = useFileChatContext();
   const [input, setInput] = useState<string>('')
+  const fileRef = useRef<HTMLInputElement>(null);
 
   // Effect hook to subscribe to pubsub events and update the message state hook
   useEffect(() => {
-    const messageCB = (evt: CustomEvent<Message>) => {
+    const messageCB = async (evt: CustomEvent<Message>) => {
       console.log('gossipsub console log', evt.detail)
       // FIXME: Why does 'from' not exist on type 'Message'?
       const { topic, data } = evt.detail
+
+      switch (topic) {
+        case CHAT_TOPIC: {
+          chatMessageCB(evt, topic, data)
+          break
+        }
+        case CHAT_FILE_TOPIC: {
+          chatFileMessageCB(evt, topic, data)
+          break
+        }
+        default: {
+          console.log(`Unexpected gossipsub topic: ${topic}`)
+          break
+        }
+      }
+    }
+
+    const chatMessageCB = (evt: CustomEvent<Message>, topic: string, data: Uint8Array) => {
       const msg = new TextDecoder().decode(data)
       console.log(`${topic}: ${msg}`)
 
       // Append signed messages, otherwise discard
       if (evt.detail.type === 'signed') {
-        setMessageHistory([...messageHistory, { msg, from: 'other', peerId: evt.detail.from.toString() }])
+        setMessageHistory([...messageHistory, { msg, fileObjectUrl: undefined, from: 'other', peerId: evt.detail.from.toString() }])
+      }
+    }
+
+    const chatFileMessageCB = async (evt: CustomEvent<Message>, topic: string, data: Uint8Array) => {
+      const fileId = new TextDecoder().decode(data)
+
+      // if the message isn't signed, discard it.
+      if (evt.detail.type !== 'signed') {
+        return
       }
+      const senderPeerId = evt.detail.from;
+
+      const stream = await libp2p.dialProtocol(senderPeerId, FILE_EXCHANGE_PROTOCOL)
+      await pipe(
+        [uint8ArrayFromString(fileId)],
+        (source) => lp.encode(source),
+        stream,
+        (source) => lp.decode(source),
+        async function(source) {
+          for await (const data of source) {
+            const body: Uint8Array = data.subarray()
+            console.log(`request_response: response received: size:${body.length}`)
+
+            const msg: ChatMessage = {
+              msg: newChatFileMessage(fileId, body),
+              fileObjectUrl: window.URL.createObjectURL(new Blob([body])),
+              from: 'other',
+              peerId: senderPeerId.toString(),
+            }
+            setMessageHistory([...messageHistory, msg])
+          }
+        }
+      )
     }
 
     libp2p.services.pubsub.addEventListener('message', messageCB)
 
+    libp2p.handle(FILE_EXCHANGE_PROTOCOL, ({ stream }) => {
+      pipe(
```

Comment:
We've recently [refactored](https://github.com/libp2p/js-libp2p/blob/b1024c6cc352b0e3bb17f7a06dff0f7d9a3e24e0/packages/libp2p/src/identify/identify.ts#L195-L211) the way we open / close streams, I would suggest this route for more predictable closing of streams.

Reply:
Thank you @maschad! Very good point. But, can I change this as next PRs because I have to upgrade the dependency `@libp2p/bootstrap` from 8.0.0 to 9.0.0 to use the refactored way that you suggested? I think it would be easier for us to draw the line with this PR first, and make this compatible with the latest version of js-libp2p.

Reply:
Sure

Reply:
Sure: https://github.com/libp2p/universal-connectivity/issues/81